src/operator/warp.cc

```diff
@@ -0,0 +1,277 @@
+        bool bottomRightIsIn = xInTopLeft+1 >= 0 && xInTopLeft+1 <= inputImages_width-1 &&
+         yInTopLeft+1 >= 0 && yInTopLeft+1 <= inputImages_height-1;
+
+        int t;
+
+        for (t=0; t < inputImages_channels; t++) {
+           double gradOutValue = gradOutput_data[gradOutputAddress + t];
+           if (topLeftIsIn) {
+              double inTopLeft = inputImages_data[inTopLeftAddress + t];
+              topLeftDotProduct += inTopLeft * gradOutValue;
+              if (!onlyGrid) gradInputImages_data[gradInputImagesTopLeftAddress + t] +=
+               xWeightTopLeft * yWeightTopLeft * gradOutValue;
+           }
+
+           if (topRightIsIn) {
+              double inTopRight = inputImages_data[inTopRightAddress + t];
+              topRightDotProduct += inTopRight * gradOutValue;
+              if (!onlyGrid) gradInputImages_data[gradInputImagesTopRightAddress + t] +=
+               (1 - xWeightTopLeft) * yWeightTopLeft * gradOutValue;
+           }
+
+           if (bottomLeftIsIn) {
+              double inBottomLeft = inputImages_data[inBottomLeftAddress + t];
+              bottomLeftDotProduct += inBottomLeft * gradOutValue;
+              if (!onlyGrid) gradInputImages_data[gradInputImagesBottomLeftAddress + t] +=
+               xWeightTopLeft * (1 - yWeightTopLeft) * gradOutValue;
+           }
+
+           if (bottomRightIsIn) {
+              double inBottomRight = inputImages_data[inBottomRightAddress + t];
+              bottomRightDotProduct += inBottomRight * gradOutValue;
+              if (!onlyGrid) gradInputImages_data[gradInputImagesBottomRightAddress + t] +=
+               (1 - xWeightTopLeft) * (1 - yWeightTopLeft) * gradOutValue;
+           }
+        }
+
+        yf = - xWeightTopLeft * topLeftDotProduct + xWeightTopLeft * bottomLeftDotProduct -
+         (1-xWeightTopLeft) * topRightDotProduct + (1-xWeightTopLeft) * bottomRightDotProduct;
+        xf = - yWeightTopLeft * topLeftDotProduct + yWeightTopLeft * topRightDotProduct -
+         (1-yWeightTopLeft) * bottomLeftDotProduct + (1-yWeightTopLeft) * bottomRightDotProduct;
+
+        gradGrids_data[b*gradGrids_strideBatch + yOut*gradGrids_strideHeight +
+         xOut*gradGrids_strideWidth + 1] = yf;
+        gradGrids_data[b*gradGrids_strideBatch + yOut*gradGrids_strideHeight +
+         xOut*gradGrids_strideWidth] = xf;
+      }
+    }
+  }
+}
+}  // namespace mshadow
+namespace mxnet {
+namespace op {
+template<>
+Operator *CreateOp<cpu>(WarpParam param) {
+  return new WarpOp<cpu>(param);
+}
+
+Operator* WarpProp::CreateOperator(Context ctx) const {
+  DO_BIND_DISPATCH(CreateOp, param_);
+}
+
+DMLC_REGISTER_PARAMETER(WarpParam);
+MXNET_REGISTER_OP_PROPERTY(Warp, WarpProp)
+.describe("According to grid, this operator warp data via bilinear interpolation. "
+"Grid has two channels : displacements of x-axis and y-axis. "
+"We assume [batch, y, x, channel] format on both input and output. \n"

```

Comment:
I also think the NHWC way in Torch is somehow strange. We may try to use the NCHW setting. The overall speed may not downgrade too much since we will need an additional transpose of the input when using NCHW.